nixos/modules/services/misc/etebase-server.nix

```diff
@@ -9,30 +9,38 @@ let
     [ etebase-server daphne ]);
 
   dbConfig = {
-    sqlite3 = ''
-      engine = django.db.backends.sqlite3
-      name = ${cfg.dataDir}/db.sqlite3
-    '';
+    sqlite3 = {
+      engine = "django.db.backends.sqlite3";
+      name = "${cfg.dataDir}/db.sqlite3";
+    };
   };
 
-  defaultConfigIni = toString (pkgs.writeText "etebase-server.ini" ''
-    [global]
-    debug = false
-    secret_file = ${if cfg.secretFile != null then cfg.secretFile else ""}
-    media_root = ${cfg.dataDir}/media
-
-    [allowed_hosts]
-    allowed_host1 = ${cfg.host}
+  iniFmt = pkgs.formats.ini {};
 
-    [database]
-    ${dbConfig."${cfg.database.type}"}
-  '');
+  defaultConfigIni = {
+    global = {
+      debug = false;
+      secret_file = if cfg.secretFile != null then cfg.secretFile else "";
+      media_root = "${cfg.dataDir}/media";
+    };
+    allowed_hosts = {
+      allowed_host1 = cfg.host;
+    };
+    database = dbConfig."${cfg.database.type}";
+  };
 
-  configIni = if cfg.customIni != null then cfg.customIni else defaultConfigIni;
+  configIni = toString (iniFmt.generate "etebase-server.ini"
+    (recursiveUpdate defaultConfigIni cfg.extraConfig));
```

Comment:
Instead of doing this `recursiveUpdate`, you could use [freeform modules](https://nixos.org/manual/nixos/stable/index.html#sec-freeform-modules) to add supoptions to `extraConfig`/`settings` with preset default values. That would also make the default configuration more transparend to users.

I wonder if it would be possible to get rid of all top-level options of this module in favor of `settings`?

Reply:
Thanks for the great feedback. I really like this configuration approach.
I've adapted the module, do you mind taking a look again?

I can't get rid of all top-level options as they're used in other parts of the configuration outside of the ini.

Reply:
I think that is a quite considerable reduction in top-level options and the ones that remain there are very reasonable to have top-level!